File: template/merge/merge-settings.js
'use strict';
const fs = require('fs');

function mergeConfig(settingsPath, templatePath) {
  try {
    const home = process.env.HOME || '';

    let settings;
    try {
      settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8'));
    } catch (e) {
      if (e.code === 'ENOENT') return [false, `File not found: ${settingsPath}`];
      return [false, `Invalid JSON in ${settingsPath}: ${e.message}`];
    }

    let templateRaw;
    try {
      templateRaw = fs.readFileSync(templatePath, 'utf8');
    } catch (e) {
      return [false, `Template not found: ${templatePath}`];
    }
    let template;
    try {
      template = JSON.parse(templateRaw.replace(/__HOME__/g, home));
    } catch (e) {
      return [false, `Invalid JSON in template after __HOME__ replacement: ${e.message}`];
    }

    let changed = false;

    // Merge hooks — add only missing hook events
    const tHooks = template.hooks || {};
    if (Object.keys(tHooks).length) {
      const eHooks = (settings.hooks && typeof settings.hooks === 'object') ? {...settings.hooks} : {};
      const mHooks = {...eHooks};
      for (const [event, hooksList] of Object.entries(tHooks)) {
        if (!(event in eHooks)) {
          mHooks[event] = hooksList;
          changed = true;
        }
      }
      if (changed) settings.hooks = mHooks;
    }

    const tAuto = template.autoMemoryEnabled !== false;
    if (settings.autoMemoryEnabled !== tAuto) {
      settings.autoMemoryEnabled = tAuto;
      changed = true;
    }

    const tEnv = template.env || {};
    const eEnv = (settings.env && typeof settings.env === 'object') ? {...settings.env} : {};
    const mEnv = {...eEnv};
    for (const [k, v] of Object.entries(tEnv)) {
      if (eEnv[k] !== v) { mEnv[k] = v; changed = true; }
    }
    settings.env = mEnv;

    if (!changed) {
      try { fs.unlinkSync(settingsPath + '.bak.merge'); } catch {}
      return [false, null]; // SKIPPED
    }

    const tmpPath = settingsPath + '.merge.tmp';
    fs.writeFileSync(tmpPath, JSON.stringify(settings, null, 2) + '\n');
    JSON.parse(fs.readFileSync(tmpPath, 'utf8'));

    const backupPath = settingsPath + '.bak.evokit';
    try { fs.unlinkSync(backupPath); } catch {}
    try {
      fs.renameSync(settingsPath, backupPath);
    } catch (e) {
      try { fs.unlinkSync(tmpPath); } catch {}
      return [false, `Cannot backup ${settingsPath}: ${e.message}`];
    }

    try {
      fs.renameSync(tmpPath, settingsPath);
    } catch (e) {
      try { fs.renameSync(backupPath, settingsPath); } catch {}
      try { fs.unlinkSync(tmpPath); } catch {}
      return [false, `Cannot write ${settingsPath}: ${e.message}`];
    }

    try { fs.unlinkSync(settingsPath + '.bak.merge'); } catch {}
    return [true, null]; // MERGED
  } catch (e) {
    return [false, `Unexpected error: ${e.message}`];
  }
}

const [changed, err] = mergeConfig(process.argv[2], process.argv[3]);
if (err) {
  console.log('ERROR');
  process.stderr.write('MERGE_ERROR: ' + err + '\n');
  process.exit(1);
} else if (changed) {
  console.log('MERGED');
  process.exit(0);
} else {
  console.log('SKIPPED');
  process.exit(0);
}

File: template/merge/merge-settings.py
import json, os, sys

def merge_config(settings_path, template_path):
    try:
        home = os.environ.get('HOME', '')
        try:
            with open(settings_path, 'r') as f:
                settings = json.load(f)
        except FileNotFoundError:
            return False, "File not found: " + settings_path
        except json.JSONDecodeError as e:
            return False, "Invalid JSON in " + settings_path + ": " + str(e)

        try:
            with open(template_path, 'r') as f:
                raw = f.read().replace('__HOME__', home)
            template = json.loads(raw)
        except FileNotFoundError:
            return False, "Template not found: " + template_path
        except json.JSONDecodeError as e:
            return False, "Invalid JSON in template after __HOME__ replacement: " + str(e)

        changed = False
        template_hooks = template.get('hooks', {})
        if template_hooks:
            existing_hooks = settings.get('hooks')
            if not isinstance(existing_hooks, dict):
                existing_hooks = {}
                changed = True
            merged_hooks = dict(existing_hooks)
            for event, hooks_list in template_hooks.items():
                if event not in existing_hooks:
                    merged_hooks[event] = hooks_list
                    changed = True
            if changed:
                settings['hooks'] = merged_hooks

        template_auto = template.get('autoMemoryEnabled', True)
        if settings.get('autoMemoryEnabled') != template_auto:
            settings['autoMemoryEnabled'] = template_auto
            changed = True

        template_env = template.get('env', {})
        current_env = settings.get('env', {})
        if not isinstance(current_env, dict):
            current_env = {}
            changed = True
        merged_env = dict(current_env)
        for k, v in template_env.items():
            if current_env.get(k) != v:
                merged_env[k] = v
                changed = True
        settings['env'] = merged_env

        if not changed:
            try:
                if os.path.exists(settings_path + '.bak.merge'):
                    os.remove(settings_path + '.bak.merge')
            except OSError:
                pass
            return False, None

        tmp_path = settings_path + '.merge.tmp'
        with open(tmp_path, 'w') as f:
            json.dump(settings, f, indent=2)
            f.write('\n')
        with open(tmp_path, 'r') as f:
            json.load(f)

        backup_path = settings_path + '.bak.evokit'
        try:
            if os.path.exists(backup_path):
                os.remove(backup_path)
            os.rename(settings_path, backup_path)
        except OSError as e:
            if os.path.exists(tmp_path):
                os.remove(tmp_path)
            return False, "Cannot backup " + settings_path + ": " + str(e)

        try:
            os.rename(tmp_path, settings_path)
        except OSError as e:
            if os.path.exists(backup_path):
                os.rename(backup_path, settings_path)
            if os.path.exists(tmp_path):
                os.remove(tmp_path)
            return False, "Cannot write " + settings_path + ": " + str(e)

        try:
            if os.path.exists(settings_path + '.bak.merge'):
                os.remove(settings_path + '.bak.merge')
        except OSError:
            pass

        return True, None
    except Exception as e:
        return False, "Unexpected error: " + str(e)

if __name__ == '__main__':
    changed, err = merge_config(sys.argv[1], sys.argv[2])
    if err:
        print('ERROR', flush=True)
        sys.stderr.write('MERGE_ERROR: ' + err + '\n')
        sys.exit(1)
    elif changed:
        print('MERGED', flush=True)
        sys.exit(0)
    else:
        print('SKIPPED', flush=True)
        sys.exit(0)
